x-pack/plugins/security_solution/public/resolver/test_utilities/simulator/index.tsx

```diff
@@ -0,0 +1,290 @@
+      resolverMiddlewareFactory(dataAccessLayer),
+      // install the spyMiddleware
+      this.spyMiddleware.middleware
+    );
+
+    // Create a redux store w/ the top level Resolver reducer and the enhancer that includes the Resolver middleware and the spyMiddleware
+    this.store = createStore(resolverReducer, middlewareEnhancer);
+
+    // Create a fake 'history' instance that Resolver will use to read and write query string values
+    this.history = createMemoryHistory();
+
+    // Used for KibanaContextProvider
+    const coreStart: CoreStart = coreMock.createStart();
+
+    // Render Resolver via the `MockResolver` component, using `enzyme`.
+    this.wrapper = mount(
+      <MockResolver
+        resolverComponentInstanceID={this.resolverComponentInstanceID}
+        history={this.history}
+        store={this.store}
+        coreStart={coreStart}
+        databaseDocumentID={databaseDocumentID}
+      />
+    );
+
+    // Update the enzyme wrapper after each state transition
+    connectEnzymeWrapperAndStore(this.store, this.wrapper);
+  }
+
+  /**
+   * Call this to console.log actions (and state). Use this to debug your tests.
+   * State and actions aren't exposed otherwise because the tests using this simulator should
+   * assert stuff about the DOM instead of internal state. Use selector/middleware/reducer
+   * unit tests to test that stuff.
+   */
+  public debugActions(): /**
+   * Optionally call this to stop debugging actions.
+   */ () => void {
+    return this.spyMiddleware.debugActions();
+  }
+
+  /**
+   * Return a promise that resolves after the `store`'s next state transition.
+   * Used by `mapStateTransisions`
+   */
+  private stateTransitioned(): Promise<void> {
+    // keep track of the resolve function of the promise that has been returned.
+    let resolveState: (() => void) | null = null;
+
+    const promise: Promise<undefined> = new Promise((resolve) => {
+      // immedatiely expose the resolve function in the outer scope. it will be resolved when the next state transition occurs.
+      resolveState = resolve;
+    });
+
+    // Subscribe to the store
+    const unsubscribe = this.store.subscribe(() => {
+      // Once a state transition occurs, unsubscribe.
+      unsubscribe();
+      // Resolve the promise. The null assertion is safe here as Promise initializers run immediately (according to spec and node/browser implementations.)
+      // NB: the state is not resolved here. Code using the simulator should not rely on state or selectors of state.
+      resolveState!();
+    });
+
+    // Return the promise that will be resolved on the next state transition, allowing code to `await` for the next state transition.
+    return promise;
+  }
+
+  /**
+   * This will yield the return value of `mapper` after each state transition. If no state transition occurs for 10 event loops in a row, this will give up.
+   */
+  public async *mapStateTransisions<R>(mapper: () => R): AsyncIterable<R> {
```

Comment:
👌

Reply:
another place where better names are needed

Reply:
Actually, one quick thing here. Spelled transitions wrong here

Reply:
NO U

Reply:
https://github.com/elastic/kibana/pull/73310/commits/1e7abdf222d01bdf188545734a525b3e266b2603

Reply:
https://github.com/elastic/kibana/pull/73310/commits/1e7abdf222d01bdf188545734a525b3e266b2603 i hope you're happy